text/0000-nested-publish.md

```diff
@@ -0,0 +1,186 @@
+- Feature Name: `nested_publish`
+- Start Date: 2023-06-30
+- RFC PR: [rust-lang/rfcs#3452](https://github.com/rust-lang/rfcs/pull/3452)
+- Rust Issue: ...
+
+# Summary
+[summary]: #summary
+
+Allow Cargo packages to be bundled within other Cargo packages when they are published (not just in unpublished workspaces).
+
+# Motivation
+[motivation]: #motivation
+
+There are a number of reasons why a Rust developer currently may feel the need to create multiple library crates, and therefore multiple Cargo packages (since one package contains at most one library crate). These multiple libraries could be:
+
+* A trait declaration and a corresponding derive macro (which must be defined in a separate proc-macro library).
+* A library that uses a build script that uses another library or binary (e.g. for precomputation or bindings generation).
+* A logically singular library broken into multiple parts to speed up compilation.
+
+Currently, developers must publish these packages separately. This has several disadvantages (see the [Rationale](#rationale-and-alternatives) section for further details):
+
+* Clutters the public view of the registry with packages not intended to be usable on their own, and which may even become obsolete as internal architecture changes.
+* Requires multiple `cargo publish` operations (this could be fixed with bulk publication) and writing public metadata for each package.
+* Can result in semver violations and thus compilation failures, due to the developer not thinking about semver compatibility within the group.
+
+This RFC will allow developers to avoid all of these inconveniences and hazards by publishing a single package.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+By default (and always, prior to this RFC's implementation):
+
+* If your package contains any sub-packages, Cargo [excludes](https://doc.rust-lang.org/cargo/reference/manifest.html#the-exclude-and-include-fields) them from the `.crate` archive file produced by `cargo package` and `cargo publish`.
+* If your package contains any non-`dev` dependencies which do not give a `version = "..."`, it cannot be published to `crates.io`.
+* If your package contains `[dev-dependencies]` which do not give a `version = "..."`, they are stripped out on publication.
+
+(By “**sub-package**” we mean a package (directory with `Cargo.toml`) which is a subdirectory of another package. We shall call the outermost such package, the package being published, the “**parent package**”.)
```

Comment:
Do we need to restrict these "sub-packages" to being in sub-directories of their parent?
1. This prevents multiple packages vendoring the same package
2. There is a preference by some ([example](https://matklad.github.io/2021/08/22/large-rust-workspaces.html)) to have flat workspaces (myself included)
3. We could even consider expanding this to other forms of dependencies like `git` dependencies

Implementation wise, we just package the crate to be within the right directory hierarchy and update the path dependencies.  This could be listed in "Future Possibilities" (if the design doesn't block it) or in Alternatives / Drawbacks (if the design blocks it)

Reply:
> This prevents multiple packages vendoring the same package

IMO, this is a feature, not a bug. I'm trying to avoid creating any code duplication that would not already exist. I can imagine the use cases otherwise, though (e.g. an author might want to use an unstable “utilities” package within several packages).

> Implementation wise, we just package the crate to be within the right directory hierarchy

This could create name conflicts with files already present in the package. How do you see addressing that? It could be just an error, but in general, I don't like this kind of namespace conflation — my experience is that it will eventually bite somebody. Though I can't currently imagine a case that couldn't be worked around, since by definition the sub-packages being packaged this way can be renamed.

If this were to be supported, I would rather see the full set of packages archived as siblings in the `.crate`, so their filenames cannot conflict.

Reply:
It would be possible to include a "utilities" package as nested within multiple other packages via symlinks, as is done currently for things like distributing `LICENSE` in multiple crates of a single workspace.

Reply:
This error makes me wonder: how would cargo handle having a nested package which has a name overlap with a (transitive) dependency? (what registry is a nested package considered to be part of?) Would it be possible to `patch` a nested package, or would you have to patch the outer package as a whole?

Reply:
> IMO, this is a feature, not a bug. I'm trying to avoid creating any code duplication that would not already exist. I can imagine the use cases otherwise, though (e.g. an author might want to use an unstable “utilities” package within several packages).

I was specifically thinking of these "utilities" style scenarios, much like C++ "headers only" libraries.  See also https://internals.rust-lang.org/t/idea-light-weight-reusable-dependencies/17099

> This could create name conflicts with files already present in the package. How do you see addressing that? It could be just an error, but in general, I don't like this kind of namespace conflation — my experience is that it will eventually bite somebody. Though I can't currently imagine a case that couldn't be worked around, since by definition the sub-packages being packaged this way can be renamed.

Since its a new feature, we can define what we want and people have to deal with it.   One simple way of doing that is putting them in a `.crates/`.  You could make the directory be `.crates/<pkg-name>` and error if there are duplicates, `.crates/<ws-root-to-pkg-root-rel-path>` with some special handling for anything that doesn't fit within that model, etc.

The own downside to making everything a sibling is it removes the "it just works" nature of this proposal as we'd need to have special handling for finding the root item of the registry package compared to normal registry packages.

Reply:
OK, having `.crates/` (or something else left open as an implementation choice) seems reasonable, especially given the [previous discussion](https://github.com/rust-lang/rfcs/pull/3452#discussion_r1248385691) where we agreed that nested packages should be selected based on dependencies or something, not just directory containment. I have not yet rewritten the RFC to address that (and may not any time soon; I'm having a bit of hand pain and using dictation for half these words, which isn't great for thinking and rewriting).

Reply:
I have now updated the RFC text to allow nested packages that don't live inside the parent package's directory.